safetensors/src/tensor.rs

```diff
@@ -48,34 +49,126 @@ impl std::fmt::Display for SafeTensorError {
 
 impl std::error::Error for SafeTensorError {}
 
-struct PreparedData<'hash, 'data> {
+struct PreparedData {
     n: u64,
     header_bytes: Vec<u8>,
-    tensors: Vec<&'hash TensorView<'data>>,
+    // tensors: Vec<&'data TensorView<'data>>,
     offset: usize,
 }
 
-fn prepare<'hash, 'data>(
-    data: &'hash HashMap<String, TensorView<'data>>,
-    data_info: &'hash Option<HashMap<String, String>>,
+/// The trait necessary to enable safetensors to serialize a tensor
+/// If you have an owned tensor like this:
+///
+/// ```rust
+/// use safetensors::tensor::{View, Dtype};
+/// use std::borrow::Cow;
+/// struct Tensor{ dtype: MyDtype, shape: Vec<usize>, data: Vec<u8>}
+///
+/// # type MyDtype = Dtype;
+/// impl<'data> View for &'data Tensor{
+///    fn dtype(&self) -> Dtype{
+///        self.dtype.into()
+///    }
+///    fn shape(&self) -> &[usize]{
+///         &self.shape
+///    }
+///    fn data(&self) -> Cow<[u8]>{
+///        (&self.data).into()
+///    }
+///    fn data_len(&self) -> usize{
+///        self.data.len()
+///    }
+/// }
+/// ```
+///
+/// For a borrowed tensor:
+///
+/// ```rust
+/// use safetensors::tensor::{View, Dtype};
+/// use std::borrow::Cow;
+/// struct Tensor<'data>{ dtype: MyDtype, shape: Vec<usize>, data: &'data[u8]}
+///
+/// # type MyDtype = Dtype;
+/// impl<'data> View for Tensor<'data>{
+///    fn dtype(&self) -> Dtype{
+///        self.dtype.into()
+///    }
+///    fn shape(&self) -> &[usize]{
+///         &self.shape
+///    }
+///    fn data(&self) -> Cow<[u8]>{
+///        self.data.into()
+///    }
+///    fn data_len(&self) -> usize{
+///        self.data.len()
+///    }
+/// }
+/// ```
+///
+/// Now if you have some unknown buffer that could be on GPU for instance,
+/// you can implement the trait to return an owned local buffer containing the data
+/// on CPU (needed to write on disk)
+/// ```rust
+/// use safetensors::tensor::{View, Dtype};
+/// use std::borrow::Cow;
+///
+/// # type MyDtype = Dtype;
+/// # type OpaqueGpu = Vec<u8>;
+/// struct Tensor{ dtype: MyDtype, shape: Vec<usize>, data: OpaqueGpu }
+///
+/// impl View for Tensor{
+///    fn dtype(&self) -> Dtype{
+///        self.dtype.into()
+///    }
+///    fn shape(&self) -> &[usize]{
+///         &self.shape
+///    }
+///    fn data(&self) -> Cow<[u8]>{
+///        // This copies data from GPU to CPU.
+///        let data: Vec<u8> = self.data.to_vec();
+///        data.into()
+///    }
+///    fn data_len(&self) -> usize{
```

Comment:
```suggestion
/// struct Tensor{ dtype: MyDtype, shape: Vec<usize>, data: Vec<u8> }
///
/// // type MyDtype = Dtype;
/// impl<'data> View for &'data Tensor {
///    fn dtype(&self) -> Dtype {
///        self.dtype.into()
///    }
///    fn shape(&self) -> &[usize] {
///         &self.shape
///    }
///    fn data(&self) -> Cow<[u8]> {
///        (&self.data).into()
///    }
///    fn data_len(&self) -> usize {
///        self.data.len()
///    }
/// }
/// ```
///
/// For a borrowed tensor:
///
/// ```rust
/// use safetensors::tensor::{View, Dtype};
/// use std::borrow::Cow;
/// struct Tensor<'data>{ dtype: MyDtype, shape: Vec<usize>, data: &'data[u8] }
///
/// // type MyDtype = Dtype;
/// impl<'data> View for Tensor<'data> {
///    fn dtype(&self) -> Dtype {
///        self.dtype.into()
///    }
///    fn shape(&self) -> &[usize] {
///         &self.shape
///    }
///    fn data(&self) -> Cow<[u8]> {
///        self.data.into()
///    }
///    fn data_len(&self) -> usize {
///        self.data.len()
///    }
/// }
/// ```
///
/// Now if you have some unknown buffer that could be on GPU for instance,
/// you can implement the trait to return an owned local buffer containing the data
/// on CPU (needed to write on disk)
/// ```rust
/// use safetensors::tensor::{View, Dtype};
/// use std::borrow::Cow;
///
/// // type MyDtype = Dtype;
/// // type OpaqueGpu = Vec<u8>;
/// struct Tensor{ dtype: MyDtype, shape: Vec<usize>, data: OpaqueGpu }
///
/// impl View for Tensor {
///    fn dtype(&self) -> Dtype {
///        self.dtype.into()
///    }
///    fn shape(&self) -> &[usize] {
///         &self.shape
///    }
///    fn data(&self) -> Cow<[u8]> {
///        // This copies data from GPU to CPU.
///        let data: Vec<u8> = self.data.to_vec();
///        data.into()
///    }
///    fn data_len(&self) -> usize {
```